### PR TITLE
Speed h3 library by using FastMath implementation for trigonometric functions

### PR DESCRIPTION
**Pull request by iverase**
Currently, one of the bottlenecks of the H3 implementation are the execution of trigonometric functions. This change propose to use a faster implementation for those function which trade speed for accuracy. After the change, all test still work and performance is greatly improved.

The FastMath implementations are forked from https://github.com/jeffhain/jafama/blob/master/src/main/java/net/jafama/FastMath.java which is released under AL2 license.

### CODE REVIEW

**Comment by iverase** [comment]
@elasticsearchmachine update branch

File: libs/h3/src/test/java/org/elasticsearch/h3/FastMathTests.java [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+    public void testAtan() {
+        doTest(
+            Math::atan,
+            FastMath::atan,
+            d -> ATAN_DELTA,
+            () -> randomDoubleBetween(Double.NEGATIVE_INFINITY, Double.POSITIVE_INFINITY, true)
+        );
+    }
+
+    public void doTest(DoubleUnaryOperator expected, DoubleUnaryOperator actual, DoubleUnaryOperator delta, DoubleSupplier supplier) {
```
Comment (craigtaverner) [inline libs/h3/src/test/java/org/elasticsearch/h3/FastMathTests.java:78]: Presumably this should be private?

**Review by craigtaverner — APPROVED** [review]
Let's hope it is accurate enough that no customers/users notice results changing in 8.7!

**Comment by iverase** [comment]
@elasticmachine update branch

**Comment by iverase** [comment]
unrelated failure: https://github.com/elastic/elasticsearch/issues/92206

@elasticmachine run elasticsearch-ci/part-2

File: libs/h3/src/main/java/org/elasticsearch/h3/FastMath.java [hunk 2]
```diff
@@ -0,0 +1,1055 @@
...
+ * Copyright (C) 1993 by Sun Microsystems, Inc. All rights reserved.
+ *
+ * Developed at SunSoft, a Sun Microsystems, Inc. business.
+ * Permission to use, copy, modify, and distribute this
+ * software is freely granted, provided that this notice
+ * is preserved.
+ * =============================================================================
+ *
+ * This code sourced from:
+ * https://github.com/yannrichet/jmathplot/blob/f25426e0ab0e68647ad2b75f577c7be050ecac86/src/main/java/org/math/plot/utils/FastMath.java
```
Comment (gtback) [inline libs/h3/src/main/java/org/elasticsearch/h3/FastMath.java:29]:
```suggestion
 * https://github.com/jeffhain/jafama/blob/d7d2a7659e96e148d827acc24cf385b872cda365/src/main/java/net/jafama/FastMath.java
```

**Review by gtback — CHANGES_REQUESTED** [review]

**Review by gtback — APPROVED** [review]
I didn't mean to "request changes", so my approval is just to prevent that from blocking a merge.

**Review by jpountz — APPROVED** [review]
Given the existing accuracy trade-off of our geo support, this change feels like it wouldn't make accuracy noticeably worse, while giving a significant performance boost. We're already using the same implementation for cos() and asin() in Lucene (via `org.apache.lucene.util.SloppyMath`) and no issues have been reported to my knowledge.

**Comment by iverase** [comment]
@elasticmachine update branch

**Comment by iverase** [comment]
open https://github.com/elastic/elasticsearch/issues/92454

@elasticmachine run elasticsearch-ci/part-3

**Comment by iverase** [comment]
@elasticmachine update branch

**Pull request merged**